mirai-core-utils/src/jvmBaseMain/kotlin/Streams.kt

```diff
@@ -13,9 +13,172 @@
+        override fun toString(): String {
+            return "CompleteToken[$error]"
+        }
+    }
+
+    private val NULL_PLACEHOLDER = Symbol("null")!!
+
+    /*
+    Implementation:
+
+    Spliterator.tryAdvance():
+        - Resume coroutine
+        (*Wait for collector.emit*)
+        - Re-Suspend flow
+        - Put response to queue
+
+        - Fire response to jdk consumer
+
+    Completion & Exception caught:
+        (* Spliterator.tryAdvance(): Resume coroutine *)
+        (* No more values or exception thrown. *)
+        (* completion called *)
+        - Put the exception or the completion token to queue
+
+        - Throw exception in Spliterator.tryAdvance() if possible
+        - Return false in Spliterator.tryAdvance()
+
+     */
+    public fun <T> Flow<T>.toStream(
+        context: CoroutineContext = EmptyCoroutineContext,
+        unintercepted: Boolean = true,
+    ): Stream<T> {
+        val spliterator = FlowSpliterator(
+            flow = this,
+            coroutineContext = context,
+            unintercepted = unintercepted,
+        )
+
+        return StreamSupport.stream(spliterator, false).onClose {
+            spliterator.cancelled = true
+            spliterator.nextStep?.resumeWithException(CancellationException())
+            spliterator.nextStep = null
+        }
+    }
+
+    private class FlowSpliterator<T>(
+        private val flow: Flow<T>,
+        private val coroutineContext: CoroutineContext,
+        unintercepted: Boolean
+    ) : AbstractSpliterator<T>(
+        Long.MAX_VALUE, Spliterator.ORDERED or Spliterator.IMMUTABLE
+    ) {
+
+        private val queue = ArrayBlockingQueue<Any?>(4)
+        private var completed = false
+
+        @JvmField
+        var cancelled = false
+
+        @JvmField
+        var nextStep: Continuation<Unit>? = run {
+            val completion = object : Continuation<Unit> {
+                override val context: CoroutineContext get() = coroutineContext
+
+                override fun resumeWith(result: Result<Unit>) {
+                    nextStep = null
+                    queue.put(CompleteToken(result.exceptionOrNull()))
+                    completed = true
+                }
+
+            }
+            if (unintercepted) {
+                return@run (suspend {
+                    flow.collect { item ->
+                        suspendCoroutineUninterceptedOrReturn<Unit> { cort ->
+                            nextStep = cort
+                            queue.put(boxValue(item))
+                            COROUTINE_SUSPENDED
+                        }
+                    }
+                }).createCoroutineUnintercepted(completion)
+            }
+
+            return@run (suspend {
+                flow.collect { item ->
+                    suspendCancellableCoroutine<Unit> { cort ->
+                        nextStep = cort
+                        queue.put(boxValue(item))
+                    }
```

Comment:
```suggestion
                    suspendCancellableCoroutine<Unit> { cont ->
                        nextStep = cont
                        queue.put(boxValue(item))
                    }
```

do you mean 'cont' for 'continuation'?